oak_functions_service/src/lookup_htbl.rs

```diff
@@ -0,0 +1,575 @@
+                ));
+            }
+        }
+    }
+}
+
+impl<'a> IntoIterator for &'a LookupHtbl {
+    type Item = (&'a [u8], &'a [u8]);
+    type IntoIter = LookupHtblIter<'a>;
+
+    // Required method
+    fn into_iter(self) -> Self::IntoIter {
+        LookupHtblIter {
+            htbl: self,
+            table_index: 0usize,
+        }
+    }
+}
+
+// Reduce maps a random value in the range [0..2^32] to [0..N], which is used to compute the index
+// into the hash table.  This is faster than computing x % M.
+fn reduce(x: u32, n: u32) -> usize {
+    (((x as u64) * (n as u64)) >> 32) as usize
+}
+
+// Read a u40 index from unaligned memory LE, and return it as a usize.
+#[inline]
+fn read_index(data: &[u8], index: usize) -> usize {
+    let mut val = [0u8; 8];
+    val[0..INDEX_SIZE].copy_from_slice(&data[index..index + INDEX_SIZE]);
+    u64::from_le_bytes(val) as usize
+}
+
+// Write a u40 index to unaligned memory LE.
+#[inline]
+fn write_index(data: &mut [u8], index: usize, value: usize) {
+    let data = &mut data[index..index + INDEX_SIZE];
+    data.copy_from_slice(&value.to_le_bytes()[0..INDEX_SIZE]);
+}
+
+// Read a compressed integer length.  Returns (length, num bytes of length) Compressed integers for
+// lengths are <2-bit num_bytes><6-bit MSBs>.  num_bytes is 0 for small integers <= 64.  Lengths
+// up to 2^30 can be represented, limiting keys and values to 1GiB each.
+#[inline]
+fn read_len(data: &[u8], index: usize) -> (usize, usize) {
+    let first_byte = data[index] as usize;
+    if first_byte < 64 {
+        return (first_byte, 1usize);
+    }
+    let num_bytes = first_byte >> 6;
+    let mut val = [0u8; 8];
+    val[0..num_bytes].copy_from_slice(&data[index + 1..index + 1 + num_bytes]);
+    let len = u64::from_le_bytes(val) as usize;
+    let len = len | ((first_byte & 0x3f) << (8 * num_bytes));
+    (len, 1 + num_bytes)
+}
+
+// Write a compressed length to data.  Return the number of bytes used to represent the length.
+#[inline]
+fn write_len(data: &mut [u8], index: usize, len: usize) -> usize {
+    if len < 64 {
+        data[index] = len as u8;
+        return 1usize;
+    }
+    let mut l = len;
+    let mut num_bytes = 1usize;
+    while l >= 64 {
+        data[index + num_bytes] = l as u8;
+        l >>= 8;
+        num_bytes += 1;
+    }
+    let first_byte = ((num_bytes - 1) << 6 | l) as u8;
+    data[index] = first_byte;
+    num_bytes
+}
+
+// Multiplication is our best fast mixing primitive, and with XOR, we have a non-linear hash.  We
+// multiply by odd numbers so it is reversible.  The upper 32 bits of the result of a 64x64->64
+// multiplication has a good mix from the lower 32 bits, but the upper 32 bits of the input do not
+// effect the result's lower 32 bits.  Therefore, multiply by v + (v >> 32), so that high and low
+// bits influence the high bits of the result.  Then, mix the upper 32 bits into the lower 32.
+#[inline]
+fn hash_u64(v: u64) -> u64 {
```

Comment:
Why not use the standard hash methods from `core::hash`?

Reply:
I added the following comment:

// We don't have to use this, but it does help us gain back some performance lost in all the bit
// manipulation.  It also reduces dependencies.  Finally, this is something we will need to do
// ourselves in any case once we start defending against side-channel attacks.  We'll want to
// include a random value in this hash, with constant-time processing.

We need to go down this road anyway, so I'm just laying a good foundation.  Hashing is another area were I have some experience (e.g. HighwayHash and TwoCats).